### ISSUE TITLE
[RFC] External post-generation classifier hook API

### ISSUE THREAD

**Issue by geodesia-ai** [vllm-project/vllm#43999]
## Motivation

Many production deployments need to attach a post-generation classifier to vLLM so that every completion is scored for safety, hallucination, audit, or A/B routing before the response is returned to the client. Today the only ways to do this are:

1. Monkey-patch `build_app()` in `vllm.entrypoints.openai.api_server` to inject a Starlette middleware that wraps the response. Fragile across vLLM minor releases and breaks `streaming=True` because the response body is already partially flushed by the time the middleware sees it.
2. Run vLLM behind a reverse proxy that re-parses the OpenAI JSON, calls the classifier, and re-serializes. Doubles the request latency and loses access to per-engine signals (prompt logprobs, KV state, etc).
3. Subclass `OpenAIServingChat` / `OpenAIServingCompletion` and override `create_chat_completion`. Survives one or two releases at most because vLLM iterates fast on those classes.

None of these survive a vLLM upgrade comfortably. The current state pushes safety vendors into vLLM-version pinning and that hurts the ecosystem.

What is missing is a small, public, opt-in hook surface where an external classifier registers a callable that receives the request, the generated text, and the engine's metadata for that request, and either returns scores (non-blocking) or returns a decision (blocking, with an optional replacement body).

## Proposed API

```python
from typing import Protocol, Optional, Any

class ScoringContext:
    request_id: str
    prompt: str
    generated_text: str
    extra_fields: dict[str, Any]   # from OpenAIBaseModel.get_extra_fields()
    finish_reason: str
    prompt_token_ids: list[int]
    output_token_ids: list[int]
    request_metadata: dict[str, Any]   # opaque per-request state

class ExternalClassifierHook(Protocol):
    name: str
    blocking: bool                  # True = decision; False = annotation only
    timeout_ms: int                 # per-call timeout

    async def score(self, ctx: ScoringContext) -> dict[str, Any]:
        """Return a dict of scores. If `blocking` is True and the dict
        contains the key 'block' with a truthy value, the response is
        replaced with `dict['replacement']` or with a default refusal."""
        ...

# Registration on LLMEngine / AsyncLLMEngine:
engine.register_external_classifier_hook(hook)

# Propagation: every classifier's output appears under
# RequestOutput.metadata["external_scores"][hook.name].
```

## Lifecycle

1. Generation completes (or finishes a token boundary, for streaming).
2. The engine builds a `ScoringContext`.
3. All registered hooks are invoked concurrently with `asyncio.gather`, each bounded by its declared timeout.
4. Hook return values are merged into `RequestOutput.metadata["external_scores"]`.
5. If any blocking hook returned a "block" decision, the response is replaced before serialization.
6. The OpenAI serving layer serializes the modified response.

For streaming responses, blocking hooks are invoked at end-of-stream and can replace the final chunk only. Hooks that want to inspect mid-stream tokens go through the (future) forward-hook API, not this one.

## Non-goals

* Mid-stream token mutation. That belongs in a separate RFC about the forward-hook API.
* Synchronous hooks. Everything is async to allow concurrent classifiers without head-of-line blocking.
* Registering hooks via env var or CLI. Registration is programmatic (`engine.register_external_classifier_hook(hook)`) and is meant to be invoked by a `vllm.general_plugins` entry-point.
* New auth/quota surface. Hooks run with the same trust level as the engine; vendors are expected to harden their own hook code.

## Alternatives considered

* Middleware-only. Cannot see prompt_token_ids, prompt logprobs, or the per-request metadata the engine keeps. Also bypassed for `/v1/completions` raw text path.
* gRPC sidecar. Adds a hop, complicates deployment, and makes the per-token APIs (logprobs) unavailable.
* RequestOutput post-processor decorator. Closer in spirit but does not address streaming or the blocking case, and forces every consumer to register at app build time.

## Open questions

1. Should `score()` be allowed to set sampling parameters for a retry? Useful for "if the first generation was flagged, regenerate at temperature=0" but it complicates the contract considerably.
2. Should hook outputs be serialized into the OpenAI response automatically (under a `metadata` field) or kept on the Python side only? My current preference is the latter so that we do not pollute the OpenAI schema.
3. Where exactly to mount the registration: on `LLMEngine`, on `AsyncLLMEngine`, on the `OpenAIServing*` classes, or all of the above? My current preference is `AsyncLLMEngine` because that is what every entrypoint constructs and it shares the lifetime of the engine.
4. Should the hook be allowed to access the KV cache? Probably no for v1, yes for a future RFC.

## Prior art

* Anthropic's "guardrails" approach, conceptually similar but server-internal.
* NeMo Guardrails, runs in front of the model; this RFC would let it run inside.
* Llama Guard, currently has to be wired manually by every vLLM user.

Happy to iterate on the shape. The minimum viable API for me is "an async callable that runs after generation and whose output ends up on `RequestOutput.metadata`". Everything else (timeouts, blocking semantics, replacement body) is negotiable.

**Comment by kantik001** [comment]
## Concrete consumer + measured overhead

We hit the same gap this RFC describes while wiring **decoded-text grounding verify** (numeric check vs retrieval context + optional PII) into a vLLM OpenAI serving path.

### What we tried / shipped outside the engine
OpenAI-compatible reverse proxy that calls an external gRPC verifier after generation (and mid-stream via text deltas):
- Adapter: https://github.com/kantik001/grounded-vllm
- Verifier: https://github.com/kantik001/grounded-guardrails (`VerifyText` / `VerifyStream` on `:50052`)

This matches alternative (2) in the RFC. It works across vLLM versions, but cannot see engine metadata (prompt token ids, logprobs, KV) and re-parses the OpenAI JSON.

### Why not a logits processor
Numeric/PII checks need the **detokenized** string. Calling gRPC from `apply()` every step would be wrong and slow. We opened a small docs clarification:
- https://github.com/vllm-project/vllm/pull/50051

### Overhead (CPU only; mock upstream, no model time)
On Windows / Zen+ / Python 3.12, 200 iters:

| Layer | p50 | p99 |
|------|----:|----:|
| Guardrails `VerifyText` | 0.38 ms | 3.34 ms |
| Proxy verify add (on − off) | **+0.71 ms** | **+0.69 ms** |

Full notes: https://github.com/kantik001/grounded-vllm/blob/main/OVERHEAD.md

### RFC API feedback (from this consumer)
1. **Blocking + replacement body** is the critical path for grounded verify (HTTP 422 / refusal), not only annotation scores.
2. **End-of-stream hook for SSE** is enough for v1; mid-stream can stay out of scope.
3. Please expose `extra_fields` / `vllm_xargs` (we pass `grounded_context` today via proxy `extra_body`) on `ScoringContext`.
4. Prefer registration via `vllm.general_plugins` entry points so vendors do not monkey-patch `build_app()`.

Happy to prototype against an experimental hook branch if maintainers want a reference consumer.

**Comment by ShauryaVM** [comment]
I'd like to implement the MVP in this RFC if maintainers are ok with the scope below. Design credit to @geodesia-ai; consumer constraints from @kantik001 (blocking + EOS + `vllm_xargs` + plugin registration, rather than a reverse proxy).

### v1 scope

**Mount on the serving layer, not `AsyncLLMEngine`.** Hooks need the detokenized string, request `vllm_xargs`, and the ability to replace the HTTP body. `vllm.general_plugins` already runs in engine-core and worker processes, so registering a classifier there would import a second model (Llama Guard, etc.) onto every GPU worker. `RequestOutput` also has no `metadata` field today.

Call site: a shared helper on `GenerateBaseServing` after the final `RequestOutput` (`finished=True`) and before OpenAI serialization. Streaming invokes the hook once at EOS; a blocking hook may replace only the last chunk / final response. Chat + completion in v1; Responses / Anthropic as follow-up.

**Frontend-only plugin group** (same posture as `vllm.stat_logger_plugins` / `vllm.endpoint_plugins`), allowlisted with `VLLM_PLUGINS`. No CLI flag. A `general_plugins` `register()` would execute in workers; a weight-loading guardrail must not.

```python
class ScoringContext:
    request_id: str
    prompt: str | None
    generated_text: str
    finish_reason: str | None
    prompt_token_ids: list[int] | None
    output_token_ids: list[int]
    vllm_xargs: dict[str, Any]  # request extra fields / extra_body

class PostGenerationHook(Protocol):
    name: str
    blocking: bool
    timeout_ms: int
    async def score(self, ctx: ScoringContext) -> dict[str, Any]: ...
```

- Annotation: merge `score()` onto a new optional `RequestOutput.metadata["external_scores"][hook.name]`. Not serialized into the OpenAI JSON in v1.
- Blocking: `{ "block": true, "replacement": ... }` swaps the response (or last SSE chunk) for `replacement` or a default refusal.
- Timeout: annotation hooks fail-open (warn + skip); blocking hooks fail-closed with the default refusal. Happy to flip this.

**Out of v1:** mid-stream mutation, resample/retry with new sampling params, KV-cache access, putting scores on the HTTP body.

### RFC open questions

1. `score()` setting sampling params for a retry — **no** for v1.
2. Serialize hook output into the OpenAI response — **no** for v1 (Python-side only).
3. Registration mount — **serving layer**, not `AsyncLLMEngine`.
4. KV cache — **no** for v1.

Tests: dummy plugin; non-stream + stream EOS; timeout; block+replace; default path unchanged when no plugin is loaded.

If this is acceptable I'll open a PR against this issue. If you'd rather keep registration on `AsyncLLMEngine` / `vllm.general_plugins` as originally written, say so and I'll follow that instead.

cc @DarkLight1337 @njhill @noooop

**Comment by DarkLight1337** [comment]
I think you could directly implement this as an endpoint plugin.

**Comment by ShauryaVM** [comment]
Thanks @DarkLight1337 — I'll implement this as an optional hook on `EndpointPlugin` (same frontend-only `VLLM_PLUGINS` gating, no new plugin group) and invoke it from the serving layer at EOS so a classifier can see `RequestOutput` without shadowing `/v1/chat/completions`. Existing plugins that only attach routes stay unchanged.

**Comment by ShauryaVM** [comment]
PR up as an optional `EndpointPlugin.post_generation` hook (no new plugin group): https://github.com/vllm-project/vllm/pull/53512
